Could not find path in NewClassTree.treePath

error: SourceChecker.typeProcess: unexpected Throwable (AssertionError) while processing ThrowableRef.java; message: Could not find path in NewClassTre.treePath=com.sun.source.util.TreePath@6aba2b86
  methodInvocation=new ThrowableRef(x instanceof CompletionException ? x : new CompletionException(x))
  Compilation unit: ThrowableRef.java
  Exception: java.lang.AssertionError: Could not find path in NewClassTre.treePath=com.sun.source.util.TreePath@6aba2b86
  methodInvocation=new ThrowableRef(x instanceof CompletionException ? x : new CompletionException(x)); Stack trace: org.checkerframework.framework.util.typeinference.TypeArgInferenceUtil.assignedTo(TypeArgInferenceUtil.java:198)
...

See [gist](https://gist.github.com/dbrosoft/307d5f76eaf3f061fbf5c077047c75e3) for source, command line, and full stack trace.

--- a/checker/tests/nullness/Issue672.java
+++ b/checker/tests/nullness/Issue672.java
@@ -1,0 +1,23 @@
+// Testcase for Issue 672
+// https://github.com/typetools/checker-framework/issues/672
+
+final class Issue672 extends Throwable {
+    final Throwable ex;
+
+    Issue672(Throwable x) { ex = x; }
+
+    static Issue672 test1(Throwable x, boolean flag) {
+        return new Issue672(x instanceof Exception ? x
+                : ((flag ? x :new Issue672(x))));
+    }
+
+    static Issue672 test2(Throwable x, boolean flag) {
+        return (new Issue672(x instanceof Exception ? x
+                : ((flag ? x :new Issue672(x)))));
+    }
+
+    static Issue672 test3(Throwable x) {
+        return test1(x instanceof Exception ? x
+                : new Issue672(x), false);
+    }
+}

--- a/checker/tests/nullness/java8/Issue557.java
+++ b/checker/tests/nullness/java8/Issue557.java
@@ -33,6 +33,15 @@ class Issue557a {
     MyOpt<String> opt2() {
       return MyOpt.empty();
     }
+
+    MyOpt<String> opt3(boolean flag) {
+      return flag ? MyOpt.of("Hello") : (flag ? MyOpt.empty() : MyOpt.empty());
+    }
+    void foo(MyOpt<String> param) {
+    }
+    void callFoo(boolean flag) {
+      foo(flag ? MyOpt.of("Hello") : (flag ? MyOpt.empty() : MyOpt.empty()));
+    }
 }
 
 class Issue557b {

--- a/framework/src/org/checkerframework/framework/util/typeinference/TypeArgInferenceUtil.java
+++ b/framework/src/org/checkerframework/framework/util/typeinference/TypeArgInferenceUtil.java
@@ -33,6 +33,7 @@ import javax.lang.model.util.Types;
 
 import com.sun.source.tree.AssignmentTree;
 import com.sun.source.tree.CompoundAssignmentTree;
+import com.sun.source.tree.ConditionalExpressionTree;
 import com.sun.source.tree.ExpressionTree;
 import com.sun.source.tree.LambdaExpressionTree;
 import com.sun.source.tree.MemberSelectTree;
@@ -147,11 +148,15 @@ public class TypeArgInferenceUtil {
             AnnotatedExecutableType method = AnnotatedTypes.asMemberOf(types, atypeFactory, receiver, methodElt);
             int treeIndex = -1;
             for (int i = 0; i < method.getParameterTypes().size(); ++i) {
-                if (TreeUtils.skipParens(methodInvocation.getArguments().get(i)) == path.getLeaf()) {
+                ExpressionTree argumentTree = methodInvocation.getArguments().get(i);
+                if (isArgument(path, argumentTree)) {
                     treeIndex = i;
                     break;
                 }
             }
+            assert treeIndex != -1 :  "Could not find path in MethodInvocationTree.\n"
+                                      + "treePath=" + path.toString() + "\n"
+                                      + "methodInvocation=" + methodInvocation;
 
             if (treeIndex == -1) {
                 return null;
@@ -188,13 +193,14 @@ public class TypeArgInferenceUtil {
             AnnotatedExecutableType constructor = AnnotatedTypes.asMemberOf(types, atypeFactory, receiver, constructorElt);
             int treeIndex = -1;
             for (int i = 0; i < constructor.getParameterTypes().size(); ++i) {
-                if (TreeUtils.skipParens(newClassTree.getArguments().get(i)) == path.getLeaf()) {
+                ExpressionTree argumentTree = newClassTree.getArguments().get(i);
+                if (isArgument(path, argumentTree)) {
                     treeIndex = i;
                     break;
                 }
             }
 
-            assert treeIndex != -1 :  "Could not find path in NewClassTre."
+            assert treeIndex != -1 :  "Could not find path in NewClassTree.\n"
                     + "treePath=" + path.toString() + "\n"
                     + "methodInvocation=" + newClassTree;
             if (treeIndex == -1) {
@@ -222,6 +228,24 @@ public class TypeArgInferenceUtil {
         ErrorReporter.errorAbort("AnnotatedTypes.assignedTo: shouldn't be here!");
         return null; // dead code
     }
+
+    /**
+     * Returns whether argumentTree is the tree at the leaf of path.
+     * if tree is a conditional expression, isArgument is called recursively on the true
+     * and false expressions.
+     */
+    private static boolean isArgument(TreePath path, ExpressionTree argumentTree) {
+        argumentTree = TreeUtils.skipParens(argumentTree);
+        if (argumentTree == path.getLeaf()) {
+            return true;
+        } else if (argumentTree.getKind() == Kind.CONDITIONAL_EXPRESSION) {
+            ConditionalExpressionTree conditionalExpressionTree = (ConditionalExpressionTree) argumentTree;
+            return isArgument(path, conditionalExpressionTree.getTrueExpression())
+                   || isArgument(path, conditionalExpressionTree.getFalseExpression());
+        }
+        return false;
+    }
+
     /**
      * If the variable's type is a type variable, return getAnnotatedTypeLhsNoTypeVarDefault(tree).
      * Rational:

--- a/javacutil/src/org/checkerframework/javacutil/TreeUtils.java
+++ b/javacutil/src/org/checkerframework/javacutil/TreeUtils.java
@@ -330,31 +330,39 @@ public final class TreeUtils {
      *   <li>VariableTree</li>
      * </ul>
      *
-     * If the leaf is a ConditionalExpressionTree, then recur on the leaf.
+     * If the leaf is a ConditionalExpressionTree or ParenthesizedTree, then recur on the leaf.
      *
      * Otherwise, null is returned.
      *
      * @return  the assignment context as described
      */
     public static Tree getAssignmentContext(final TreePath treePath) {
-        TreePath path = treePath.getParentPath();
+        TreePath parentPath = treePath.getParentPath();
 
-        if (path == null) {
+        if (parentPath == null) {
             return null;
         }
-        Tree node = path.getLeaf();
-        if ((node instanceof AssignmentTree) ||
-                (node instanceof CompoundAssignmentTree) ||
-                (node instanceof MethodInvocationTree) ||
-                (node instanceof NewArrayTree) ||
-                (node instanceof NewClassTree) ||
-                (node instanceof ReturnTree) ||
-                (node instanceof VariableTree)) {
-            return node;
-        } else if (node instanceof ConditionalExpressionTree) {
-            return getAssignmentContext(path);
+
+        Tree parent = parentPath.getLeaf();
+        switch (parent.getKind()) {
+        case PARENTHESIZED:
+        case CONDITIONAL_EXPRESSION:
+            return getAssignmentContext(parentPath);
+        case ASSIGNMENT:
+        case METHOD_INVOCATION:
+        case NEW_ARRAY:
+        case NEW_CLASS:
+        case RETURN:
+        case VARIABLE:
+            return parent;
+        default:
+            // 11 Tree.Kinds are CompoundAssignmentTrees,
+            // so use instanceof rather than listing all 11.
+            if (parent instanceof CompoundAssignmentTree) {
+                return parent;
+            }
+            return null;
         }
-        return null;
     }
 
     /**

